src/appleseed/renderer/modeling/bssrdf/directionaldipolebssrdf.cpp

```diff
@@ -79,10 +103,230 @@ namespace
+
+                sum_alpha_prime += alpha_prime;
+                values->m_cdf[i] = sum_alpha_prime;
+                const double mfp = static_cast<double>(values->m_mean_free_path[i]);
+                values->m_max_mean_free_path = std::max(values->m_max_mean_free_path, mfp);
+
+                const double sigma_tr = 1.0 / mfp;
+                const double sigma_t_prime = sigma_tr / std::sqrt( 3.0 * (1.0 - alpha_prime));
+
+                values->sigma_s_prime()[i] = alpha_prime * sigma_t_prime;
+                values->sigma_a()[i] = sigma_t_prime - values->sigma_s_prime()[i];
+            }
+
+            if (sum_alpha_prime > 0.0)
+                values->m_cdf *= static_cast<float>(1.0 / sum_alpha_prime);
+        }
+
+        virtual void evaluate(
+            const void*                 data,
+            const Vector3d&             outgoing_point,
+            const Vector3d&             outgoing_normal,
+            const Vector3d&             outgoing_dir,
+            const Vector3d&             incoming_point,
+            const Vector3d&             incoming_normal,
+            const Vector3d&             incoming_dir,
+            Spectrum&                   value) const APPLESEED_OVERRIDE
+        {
+            const DirectionalDipoleBSSRDFInputValues* values =
+                reinterpret_cast<const DirectionalDipoleBSSRDFInputValues*>(data);
+
+            value.resize(values->sigma_a().size());
+            value.set(0.0f);
+
+            bssrdf(
+                values,
+                incoming_point,
+                incoming_normal,
+                incoming_dir,
+                outgoing_point,
+                outgoing_normal,
+                value);
+
+#if 0
+            // Reciprocal evaluation with the reciprocity hack. Not sure we want it.
+            bssrdf(
+                values,
+                outgoing_point,
+                outgoing_normal,
+                outgoing_dir,
+                incoming_point,
+                incoming_normal,
+                value);
+
+            value *= 0.5f;
+#endif
+        }
+      private:
+
+        static double compute_rd(double alpha_prime, double two_c1, double three_c2)
+        {
+            double cphi = 0.25 * (1.0 - two_c1);
+            double ce = 0.5 * (1.0 - three_c2);
+            double mu_tr_D = std::sqrt((1.0 - alpha_prime) * (2.0 - alpha_prime) / 3.0);
+            double myexp = std::exp(-((1.0 + three_c2) / cphi) * mu_tr_D);
+            return 0.5 * square(alpha_prime) * std::exp(-std::sqrt(3.0 * (1.0 - alpha_prime) / (2.0 - alpha_prime))) * (ce * (1.0 + myexp) + cphi / mu_tr_D * (1.0 - myexp));
+        }
+
+        static double compute_alpha_prime(double rd, double c1, double c2)
+        {
+            const double c12 = 2.0 * c1;
+            const double c23 = 3.0 * c2;
+
+            double x0 = 0.0, x1 = 1.0, xmid;
+
+            // For now simple bisection.
```

Comment:
We have a solver based on Newton's method in `foundation/math/root.h`, if that can be useful.